Improve handling of key material in KMS

tombentley opened the pull request:
### Type of change

- Refactoring

### Description

These changes seek to minimise the number of copies of key material held in proxy process memory, and to not be so reliant on timely GC to clean up key material that is no longer referenced. Obviously it can't prevent the key material from existing in the plain within the proxy process, but we can seek to manage this material more carefully.

* Move the `DestroyableRawSecretKey` to the `kms` module, so it can be used by KMS implementations
* Use `DestroyableRawSecretKey` in the `Kms` interface and fix existing KMSes and tests.
* Avoid creating a String containing the plaintext key material in the Hashicorp impl by having Jackson convert the Base64 data to a `byte[]` which we then wrap in a `DestroyableRawSecretKey`. 
* Additionally, zero the HTTP buffer after Jackson has done the deserialization.

k-wall commented on .idea/misc.xml:
```diff
@@ -6,9 +6,9 @@
       <list>
         <option value="$PROJECT_DIR$/pom.xml" />
         <option value="$PROJECT_DIR$/kroxylicious-systemtests/pom.xml" />
+        <option value="$PROJECT_DIR$/performance-tests/dependency-reduced-pom.xml" />
```
I suspect your tree  has some stale files.  `performance-tests` is no longer a Maven module so there shouldn't be a dependency-reduced-pom.xml inside it..

k-wall commented on kroxylicious-kms/src/main/java/io/kroxylicious/kms/service/DestroyableRawSecretKey.java:
```diff
@@ -0,0 +1,136 @@
+    private final byte[] key;
+
+    private DestroyableRawSecretKey(String algorithm, byte[] bytes) {
+        this.algorithm = Objects.requireNonNull(algorithm).toLowerCase(Locale.ENGLISH);
```
We've been favouring Locale.ROOT else in the code.

k-wall commented on kroxylicious-kms/src/main/java/io/kroxylicious/kms/service/DestroyableRawSecretKey.java:
```diff
@@ -0,0 +1,136 @@
+    @Override
+    public int hashCode() {
+        int retval = 0;
+        for (int i = 1; i < this.key.length; i++) {
```
```suggestion
        for (int i = 0; i < this.key.length; i++) {
```

k-wall commented on kroxylicious-kms/src/main/java/io/kroxylicious/kms/service/DestroyableRawSecretKey.java:
```diff
@@ -0,0 +1,136 @@
+        for (int i = 1; i < this.key.length; i++) {
+            retval += this.key[i] * i;
+        }
+        return retval ^ this.algorithm.hashCode();
```
I don't understand why you haven't used  `Arrays.hashCode(key);`

k-wall replied:
hashCode should include the destroyed flag.

k-wall commented on kroxylicious-kms/src/main/java/io/kroxylicious/kms/service/DestroyableRawSecretKey.java:
```diff
@@ -0,0 +1,136 @@
+    private boolean destroyed = false;
+    private final byte[] key;
+
+    private DestroyableRawSecretKey(String algorithm, byte[] bytes) {
```
```suggestion
    private DestroyableRawSecretKey(@NonNull String algorithm, @NonNull byte[] bytes) {
```

k-wall commented:
> Use DestroyableRawSecretKey in the Kms interface and fix existing KMSes and tests.

Can you tell me more the motivation for this part of the change?  We tend to prefer APIs defined in terms of  interfaces rather than concrete implementations so seeing DestroyableRawSecretKey (class) across the API saddens me.

Couldn't the API have remained in terms SecretKey?   Kms interfaces would have the obligation to return `SecretKeys` that are destroyable/support encoded, but exactly how they choose to do that is a choice for them.  Our Kms impls would use DestroyableRawSecretKey.

k-wall commented on kroxylicious-kms/src/test/java/io/kroxylicious/kms/service/DestroyableRawSecretKeyTest.java:
```diff
@@ -0,0 +1,67 @@
+class DestroyableRawSecretKeyTest {
+
+    @Test
+    void testDestroy() {
```
nit: we've be favouring omitting the test prefix to test method names.

tombentley commented on kroxylicious-kms/src/main/java/io/kroxylicious/kms/service/DestroyableRawSecretKey.java:
```diff
@@ -0,0 +1,136 @@
+ * (unlike {@link javax.crypto.spec.SecretKeySpec}).
+ */
+@NotThreadSafe
+public final class DestroyableRawSecretKey implements SecretKey {
```
`getEncoded` is defined on `java.security.Key`, so _all_ keys are encodable some how. So I don't think there's really anything to be gained by making encodable part of the class name. We do currently have 'raw' in there, which is just mimicing what `SecretKeySpec` does. I could live with `DestroyableSecretKeySpec`, I guess.

k-wall replied:
Not all SecretKeys have an encoded form.  The javadoc for java.security.Key#getEncoded:

>Returns the key in its primary encoding format, or null if this key does not support encoding.
> Returns:
> the encoded key, or null if the key does not support encoding.

Currently `io.kroxylicious.kms.service.DestroyableRawSecretKey#toDestroyableKey` accepts any SecretKey that has an encoded form.   The Objects.requireNotNull will reject any `SecretKey` where `getEncoded() == null`.  That was what prompted the idea for `Encodable` to be part of the name rather than the word `Raw`.  I thought that was more precise.

However, perhaps the real issue is the fact that `DestroyableRawSecretKey#toDestroyableKey()` will accept keys with encoding format other than `RAW`.  Either DestroyableRawSecretKey#getFormat should return the actual format of the wrapped key, or `DestroyableRawSecretKey#toDestroyableKey()` should insist that the source SecretKey has type `RAW`.  In the latter case, the class name is accurate.

tombentley commented:
> Couldn't the API have remained in terms SecretKey? Kms interfaces would have the obligation to return `SecretKeys` that are destroyable/support encoded, but exactly how they choose to do that is a choice for them. Our Kms impls would use DestroyableRawSecretKey.

If it was purely _my_ decision I think I'd stick with using `DestroyableRawSecretKey`, but I've reverted this part. It's a bit of a compromise. 

* From the PoV of the API contract, `SecretKey` is necessary but not sufficient for the guarantee I'd like to make: That all our keys are destroyable and get destroyed as soon as they're no longer needed (which is determined by the caching configuration).

* In terms of implementation it means we now end up copying the `SecretKey` provided by the `Kms` into a `DestroyableRawSecretKey` in the `DekManager`, where before we could just _use_ it. It also means we're using the `toDestroyableKey()` method in the production code (before it was only tests), which means we now ought to log if a key `destroy()` failed (because that particular KMS didn't follow our advice; this just wasn't possible before). In turn this means we grow a dependency from the `kms` module to slf4j.

None of these things are deal-breakers, but I think the trade off is more nuanced than 'interface is better than class'.

I also removed the overriding of `equals()` and `hashCode()`. They were only used in the tests and the implementation was a bit icky: Calling `destroy()` would change the `hashCode()`, potentially breaking use of `DestroyableRawSecretKey` as `HashMap` keys.

k-wall commented:
> > Couldn't the API have remained in terms SecretKey? Kms interfaces would have the obligation to return `SecretKeys` that are destroyable/support encoded, but exactly how they choose to do that is a choice for them. Our Kms impls would use DestroyableRawSecretKey.
> 
> If it was purely _my_ decision I think I'd stick with using `DestroyableRawSecretKey`, but I've reverted this part. It's a bit of a compromise.
> 
> * From the PoV of the API contract, `SecretKey` is necessary but not sufficient for the guarantee I'd like to make: That all our keys are destroyable and get destroyed as soon as they're no longer needed (which is determined by the caching configuration).
> * In terms of implementation it means we now end up copying the `SecretKey` provided by the `Kms` into a `DestroyableRawSecretKey` in the `DekManager`, where before we could just _use_ it. It also means we're using the `toDestroyableKey()` method in the production code (before it was only tests), which means we now ought to log if a key `destroy()` failed (because that particular KMS didn't follow our advice; this just wasn't possible before). In turn this means we grow a dependency from the `kms` module to slf4j.
> 
> None of these things are deal-breakers, but I think the trade off is more nuanced than 'interface is better than class'.
> 
> I also removed the overriding of `equals()` and `hashCode()`. They were only used in the tests and the implementation was a bit icky: Calling `destroy()` would change the `hashCode()`, potentially breaking use of `DestroyableRawSecretKey` as `HashMap` keys.

Thanks for the explanation. On balance, I have preference for the way the code is now, but if others feel strongly for DestroyableRawSecretKey being the API fine.

SamBarker commented on kroxylicious-kms/src/main/java/io/kroxylicious/kms/service/DestroyableRawSecretKey.java:
```diff
@@ -0,0 +1,138 @@
+            source.destroy();
+        }
+        catch (DestroyFailedException e) {
+            LOGGER.warn("Failed to destroy key of {}", source.getClass());
```
Should we log the destroy failed exception? Or is that effectively a proxy for unsupported operation?

tombentley replied:
We can't throw if we want this method to work with `SecretKeySpec`, which is returned by other parts of JCE and which 3rd party implementers of `Kms` are likely to be using. And if we do throw, we are in a similar situation to when we were using DRSK in the `Kms` interface, i.e, you _have_ to be using a `SecreyKey` that is actually destroyable or it won't work. Only doing it this way means it's a runtime exception, rather than an compile time error.

If that's our requirement I'd prefer we reflected it in a type. We could do that, I suppose, by subclassing `SecretKey` to remove the `throws DestroyFailedException`.

SamBarker replied:
To be clear I wasn't objecting to what it currently does, I was just wondering if we should include the exception we caught in the logging.

SamBarker commented on kroxylicious-kms/src/main/java/io/kroxylicious/kms/service/DestroyableRawSecretKey.java:
```diff
@@ -0,0 +1,138 @@
+        if (thisKey == thatKey) {
+            return true;
+        }
+        Objects.requireNonNull(thisKey).checkNotDestroyed();
+        Objects.requireNonNull(thatKey).checkNotDestroyed();
```
Feels a bit odd for these to throw rather than return false especially when this is being consumed in lambdas. Is there a reason not to include `isDestoryed` in the return condition?

tombentley replied:
Anything that looks like an equality operation gets very slippery if the operands have some state which on which the equality depends but which can change. In this instance 

* should `same(foo, foo)` return true or false for a destroyed `foo`? Returning false is seems weird, but we can make it return true by using `==`. 
* But then what about `(foo, bar)` when `foo` and `bar` are distinct objects that _had_ the same bytes? We can't make that return true. 

However, your comment prompted me to move this to become a method of the test class, rather than adding a method to the production code which has these issues.

SamBarker reviewed (commented):
The change largely looks good to me, a few questions around naming but other wise looks good

tombentley commented on kroxylicious-kms/src/main/java/io/kroxylicious/kms/service/DestroyableRawSecretKey.java:
```diff
@@ -0,0 +1,138 @@
+     * @param bytes The key material
+     * @return The new key
+     */
+    public static @NonNull DestroyableRawSecretKey byOwnershipTransfer(@NonNull String algorithm, @NonNull byte[] bytes) {
```
`takeOwnershipOf`?

SamBarker replied:
Yeah works for me.

tombentley commented on kroxylicious-kms/src/main/java/io/kroxylicious/kms/service/DestroyableRawSecretKey.java:
```diff
@@ -0,0 +1,138 @@
+     * @param bytes The key material
+     * @return The new key
+     */
+    public static @NonNull DestroyableRawSecretKey byClone(@NonNull String algorithm, @NonNull byte[] bytes) {
```
`takeCopyOf`?

SamBarker reviewed (approved):
LGTM

Thanks @tombentley

k-wall reviewed (approved):
lgtm

tombentley commented on kroxylicious-kms-test-support/src/main/java/io/kroxylicious/kms/service/SecretKeyUtils.java:
```diff
@@ -0,0 +1,30 @@
+     * @param thatKey The other key
+     * @return true if they keys have the same algorithm and key material.
+     */
+    public static boolean same(@NonNull DestroyableRawSecretKey thisKey, @NonNull DestroyableRawSecretKey thatKey) {
```
Not in this PR
